arm64: dts: qcom: Add SDM660 SoC support

minlexx opened the pull request:
Initial device tree support for Qualcomm SDM660 SoC.

SDM660 is based off MSM8998 and uses some of its drivers.
SDM630/636 are based off SDM660 SoC and they are pin
and software compatible.

The device tree is based on the CAF 4.4 kernel tree.

Features:
 * CPU nodes
 * Timer nodes
 * Interrupt controller
 * Global Clock Controller
 * Top Level Mode Multiplexer (pin controller)
 * UART node

This is inspired by and based on the work of Craig Tatlor in
https://patchwork.kernel.org/patch/10563667/

Signed-off-by: Craig Tatlor <ctatlor97@gmail.com>
Signed-off-by: Alexey Minnekhanov  <alexey.min@gmail.com>

stephan-gh commented on arch/arm64/boot/dts/qcom/sdm660-xiaomi-lavender.dts:
```diff
@@ -0,0 +1,75 @@
+
+/ {
+	model = "Xiaomi Redmi Note 7";
+	compatible = "xiaomi,lavender", "qcom,sdm660-mtp", "qcom,sdm660", "qcom,mtp";
```
MTP is some sort of reference board from Qualcomm. Downstream kernels usually keep the `mtp` compatibles here, even though the devices have nothing to do with the original MTP board. I would drop both `mtp` compatibles:

```suggestion
	compatible = "xiaomi,lavender", "qcom,sdm660";
```

minlexx replied:
:ok:

stephan-gh commented on arch/arm64/boot/dts/qcom/sdm660-xiaomi-lavender.dts:
```diff
@@ -0,0 +1,75 @@
+
+		stdout-path = "serial0:115200n8";
+
+		framebuffer@0x9d400000 {
```
Up to you, but I would probably leave that out for upstream. It's only needed as long you don't have DRM working.

minlexx replied:
:ok: Agreed, the smaller patch is the better

framebuffer_region from reserved memory is also to be removed then

stephan-gh commented on arch/arm64/boot/dts/qcom/sdm660-xiaomi-lavender.dts:
```diff
@@ -0,0 +1,75 @@
+			alignment = <0 0x400000>;
+			size = <0 0x2c00000>;
+			linux,cma-default;
+		};
```
What does this do? It doesn't seem to be defined that way for any other devices.

minlexx replied:
This seems to be the same as cma= kernel parameter of kconfig about default cma region size (it only defines size of the region, not its address). This was copied fro downstream and works, but seems to be not needed. Will remove

:ok:

stephan-gh commented on arch/arm64/boot/dts/qcom/sdm660.dtsi:
```diff
@@ -0,0 +1,378 @@
+
+/ {
+	model = "Qualcomm Technologies, Inc. SDM 660";
+	compatible = "qcom,sdm660-mtp", "qcom,sdm660", "qcom,mtp";
```
These two should be always set by the device-specific device tree part. There is no need to have this here.

minlexx replied:
:ok:

stephan-gh commented on arch/arm64/boot/dts/qcom/sdm660.dtsi:
```diff
@@ -0,0 +1,378 @@
+			compatible = "arm,armv8";
+			reg = <0x0 0x100>;
+			enable-method = "psci";
+			efficiency = <1126>;
```
AFAICT this is not a property supported by mainline. I'm not familiar with big.LITTLE, maybe check how similar SoCs define it.

minlexx replied:
Looks like it does not exist and I can't find anything similar. Wait, there is https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/arm/cpu-capacity.txt . Need to read

minlexx replied:
Also arm,armv8 compatibe is listed in https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/arm/cpus.yaml#n118 as `arm,armv8 # Only for s/w models`. Mybe it should be `qcom,kryo`? Internet says that sdm660 has Kryo 260.

minlexx replied:
https://www.qualcomm.com/products/snapdragon-630-mobile-platform down below on this page there is a comparsion of sdm630 vs sdm660. I see sdm660 using Kryo and sdm630 cortex-a53 (and another difference is adreno508 vs adreno512)

stephan-gh replied:
I think the correct way to do would be to define a new `qcom,kryo260` compatible, similar to `qcom,kryo385`. The unversioned `qcom,kryo` compatible is likely only intended for MSM8996, which was the first SoC with Kryo cores. Back then they were not versioned apparently.

You can see this if you scroll down to the comparison on https://www.qualcomm.com/products/snapdragon-820-mobile-platform:

- Snapdragon 820 Mobile Platform has a Qualcomm® Kryo™ CPU
- Snapdragon 835 Mobile Platform has a Qualcomm® Kryo™ **280** CPU

However, the compatibles are not used very consistently. For example, msm8998 uses the generic `arm,armv8` compatible, which shouldn't be used for HW cores as you mentioned.

minlexx replied:
I'll try to do it (new compatible) and finish final split into 3 commits tomorrow

stephan-gh commented on arch/arm64/boot/dts/qcom/sdm660.dtsi:
```diff
@@ -0,0 +1,378 @@
+
+		intc: interrupt-controller@17a00000 {
+			compatible = "arm,gic-v3";
+			// ^^ sdm630 should propbably add "qcom,msm8996-gic-v3"
```
It seems like `qcom,msm8996-gic-v3` enables a quirk for broken firmware on MSM8996. Unless you are having problems I guess that you don't need it?

minlexx replied:
I was initially testing with `"qcom,msm8996-gic-v3", "arm,gic-v3"` because I saw this commit by opendata: https://gitlab.com/opendata26/linux/commit/f3a3a809f141ec26e3533ff9cf3eeecef8893cc6

Look closely, he places `return` exactly where in current kernel there is a check for msm8996 broken firmware quirk, before `rbase = gic_data_rdist_rd_base();` : https://elixir.bootlin.com/linux/v5.5-rc7/source/drivers/irqchip/irq-gic-v3.c#L219

But later when I had reliable UART output I re-tested with bare "arm,gic-v3" and had no problems so far.. So, opendata was porting sdm630 and I assumed that SoC may have some problems with broken firmware? Or maybe it is device dependent.. Remove this comment?

stephan-gh replied:
Qcom may have fixed the problem on newer SoCs. The firmware of SDM660 may not be similar to SDM630 at all, we don't know. I would just remove the comment if you don't notice any problem.

minlexx replied:
:ok:

stephan-gh commented on arch/arm64/boot/dts/qcom/sdm660.dtsi:
```diff
@@ -0,0 +1,378 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/* Copyright (c) 2018, Craig Tatlor.
```
I know this is used that way in some files, but personally this comment format is hurting my eyes :D

```suggestion
/*
 * Copyright (c) 2018, Craig Tatlor.
```

minlexx replied:
:) :ok:

stephan-gh commented:
Also, I would split it into two commits:
- add sdm660
- add xiamoi-lavender

stephan-gh commented on arch/arm64/boot/dts/qcom/sdm660.dtsi:
```diff
@@ -0,0 +1,376 @@
+			clock-output-names = "xo_board";
+		};
+
+		sleep_clk {
```
You'll need labels for these two clocks if you want to use them later.

e.g. MSM8996 has:

```c
xo_board: xo_board { /* ... */ };
sleep_clk: sleep_clk { /* ... */ };
```

minlexx replied:
I'm not sure we can introduce labels for things that are not referenced right now?

stephan-gh replied:
I guess you can add them later. On the other hand, I don't think it will produce warnings. Shared device tree fragments often have many labels, but only few are usually used in the device-specific device tree part.

stephan-gh commented on arch/arm64/boot/dts/qcom/sdm660.dtsi:
```diff
@@ -0,0 +1,376 @@
+			capacity-dmips-mhz = <1024>;
+			next-level-cache = <&L2_1>;
+			L2_1: l2-cache {
+				compatible = "arm,arch-cache";
```
I think this should be

```suggestion
				compatible = "cache";
```

msm8998 and downstream are the only users of the `arm,arch-cache` compatible - it does not seem to be referenced or documented anywhere, so I'm guessing this is a "leak" from downstream that made its way into mainline for msm8998.

minlexx replied:
Seems like it, :ok:

stephan-gh commented on arch/arm64/boot/dts/qcom/sdm660-xiaomi-lavender.dts:
```diff
@@ -0,0 +1,53 @@
+	chosen {
+		#address-cells = <2>;
+		#size-cells = <2>;
+		ranges;
```
You can remove these 3 properties if you are not using simple-framebuffer

minlexx replied:
:ok:

stephan-gh commented on arch/arm64/boot/dts/qcom/sdm660-xiaomi-lavender.dts:
```diff
@@ -0,0 +1,49 @@
+	qcom,board-id = <0x20008 0>;
+	qcom,pmic-id = <0x0001001b 0x0101011a 0x0 0x0>,
+			<0x0001001b 0x0201011a 0x0 0x0>,
+			<0x0001001b 0x0102001a 0x0 0x0>;
```
Did you check if these are necessary?

minlexx replied:
I did, it boots without all those IDs. They are probably necessary for DTBs in dtbo partition though.. I'm not going to check that now

minlexx replied:
When booting twrp
```
Decompressing kernel image start: 3945 ms
Decompressing kernel image done: 4416 ms
Dtbo hdr magic mismatch 0, with D7B7AB1E
Best match DTB tags 317/00020008/0x00000000/0/1001B/102001A/0/0/(offset)0x979E72A8/(size)0x00084890
Using pmic info 0x1001B/0x102001A/0x0/0x0 for device 0x2001B/0x102001A/0x0/0x0
Memory Base Address: 0x40000000
smem protocol = 9C2FB0A0
```
When booting my image
```
Memory Base Address: 0x40000000
Decompressing kernel image start: 2648 ms
Decompressing kernel image done: 2800 ms
Dtbo hdr magic mismatch 0, with D7B7AB1E
DTB offset is incorrect, kernel image does not have appended DTB
smem protocol = 9C2FF0A0
```

minlexx replied:
IDs removed, can be booted without them

minlexx committed:
dt-bindings: arm: Add kryo260 compatible

Kryo260 is found in SDM660, so add it in list of cpu compatibles

Signed-off-by: Alexey Minnekhanov <alexey.min@gmail.com>

minlexx committed:
arm64: dts: qcom: Add SDM660 SoC support

Initial device tree support for Qualcomm SDM660 SoC.

SDM660 is based off MSM8998 and uses some of its drivers.
SDM630/636 are based off SDM660 SoC and they are pin
and software compatible.

The device tree is based on the CAF 4.4 kernel tree.

Features:
 * CPU nodes
 * Timer nodes
 * Interrupt controller
 * Global Clock Controller
 * Top Level Mode Multiplexer (pin controller)
 * UART node

This is inspired by and based on the work of Craig Tatlor in
https://patchwork.kernel.org/patch/10563667/

Signed-off-by: Craig Tatlor <ctatlor97@gmail.com>
Signed-off-by: Alexey Minnekhanov <alexey.min@gmail.com>

minlexx committed:
arm64: dts: qcom: Add Xiaomi Redmi Note 7 (lavender)

This adds the initial device tree support for Xiaomi
Redmi Note 7 (lavender) phone. It is based on SDM660
SoC. Currently it can be booted into initrd with a
shell over UART and you can get kernel boot logs
from a pstore-ramoops.

Signed-off-by: Alexey Minnekhanov <alexey.min@gmail.com>

minlexx commented:
Sooo now this MR is approaching its final form... everything split and tidied up, ready to review :)

minlexx commented:
```
~/d/k/linux (qcom-sdm660-upstreaming|✔) $ scripts/checkpatch.pl --patch ./0001-dt-bindings-arm-Add-kryo260-compatible.patch
total: 0 errors, 0 warnings, 7 lines checked

./0001-dt-bindings-arm-Add-kryo260-compatible.patch has no obvious style problems and is ready for submission.
```

```
~/d/k/linux (qcom-sdm660-upstreaming|✔) $ scripts/checkpatch.pl --patch ./0002-arm64-dts-qcom-Add-SDM660-SoC-support.patch
Traceback (most recent call last):
  File "scripts/spdxcheck.py", line 10, in <module>
    import git
ModuleNotFoundError: No module named 'git'
WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
#33: 
new file mode 100644

total: 0 errors, 1 warnings, 375 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

./0002-arm64-dts-qcom-Add-SDM660-SoC-support.patch has style problems, please review.

NOTE: If any of the errors are false positives, please report
      them to the maintainer, see CHECKPATCH in MAINTAINERS.
```

```
~/d/k/linux (qcom-sdm660-upstreaming|✔) [1] $ scripts/checkpatch.pl --patch ./0003-arm64-dts-qcom-Add-Xiaomi-Redmi-Note-7-lavender.patch
Traceback (most recent call last):
  File "scripts/spdxcheck.py", line 10, in <module>
    import git
ModuleNotFoundError: No module named 'git'
WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
#29: 
new file mode 100644

WARNING: DT compatible string "xiaomi,lavender" appears un-documented -- check ./Documentation/devicetree/bindings/
#46: FILE: arch/arm64/boot/dts/qcom/sdm660-xiaomi-lavender.dts:13:
+       compatible = "xiaomi,lavender", "qcom,sdm660";

WARNING: DT compatible string vendor "xiaomi" appears un-documented -- check ./Documentation/devicetree/bindings/vendor-prefixes.yaml
#46: FILE: arch/arm64/boot/dts/qcom/sdm660-xiaomi-lavender.dts:13:
+       compatible = "xiaomi,lavender", "qcom,sdm660";
```

stephan-gh commented:
Regarding the last warning (DT compatible string "xiaomi,lavender" appears un-documented) - you can add a commit like https://github.com/torvalds/linux/commit/b72160fa886d757611b1c7f1127d5585fce43af2 to solve it (but again, often this warning is ignored).

stephan-gh commented on arch/arm64/boot/dts/qcom/sdm660-xiaomi-lavender.dts:
```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Copyright (c) 2020, The Linux Foundation. All rights reserved.
```
I would usually copy this from something I used as a reference and keep the date as-is. Unless you are involved with the Linux foundation. Otherwise it implies (to me at least) that the Linux Foundation has worked on it in 2020.

minlexx replied:
Should I just delete mention of Linux Foundation? This file was written by me from scratch :)

stephan-gh replied:
I guess so, yes. (IANAL)

minlexx replied:
:ok:

minlexx commented on arch/arm64/boot/dts/qcom/sdm660.dtsi:
```diff
@@ -0,0 +1,375 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Copyright (c) 2018, Craig Tatlor.
+ * Copyright (c) 2020, The Linux Foundation. All rights reserved.
```
This line was copied from opendata's original MR, (he probably copied it from downstream), but I changed date to here 2020 too, because I thought that's kind of "release date"..

stephan-gh replied:
IANAL, I would keep the original date.

minlexx replied:
:ok:

minlexx commented:
> Regarding the last warning (DT compatible string "xiaomi,lavender" appears un-documented) - you can add a commit like [torvalds/linux@b72160f](https://github.com/torvalds/linux/commit/b72160fa886d757611b1c7f1127d5585fce43af2) to solve it (but again, often this warning is ignored).

And I thought this was about https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/vendor-prefixes.yaml . Really, no Xiaomi there :(

stephan-gh commented:
> And I thought this was about https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/vendor-prefixes.yaml . Really, no Xiaomi there :(

The beauty of device tree bindings. Then you would need yet another commit to add Xiaomi there :D

minlexx commented:
I was reading through [all opendata's patches](https://patchwork.kernel.org/project/linux-arm-msm/list/?series=&submitter=176691&state=*&q=&archive=both&delegate=) (at least those that got into linux-arm-msm list)  and found this: https://patchwork.kernel.org/patch/10616213/ . I was mainly looking at tags, there is a `From:` and some `Co-developed-by:`. I guess he found original authors from qualcomm sources and mentioned them? 

Here is another patch - https://patchwork.kernel.org/patch/10563133/ . And original author complained...

![изображение](https://user-images.githubusercontent.com/13084776/73795557-f306ff80-47bb-11ea-8276-97bab48ed2c1.png)


Because I'm basically taking opendata's work and re-submitting it with minor fixes (and renaming sdm630.dtsi to sdm660.dtsi) maybe I should also add From: and Co-developed-by: for dtsi patch?..

minlexx commented:
TODO: ...and re-check the reviews from Bjorn on original v1 patch https://patchwork.kernel.org/patch/10563527/

stephan-gh commented:
> Because I'm basically taking opendata's work and re-submitting it with minor fixes (and renaming sdm630.dtsi to sdm660.dtsi) maybe I should also add From: and Co-developed-by: for dtsi patch?..

I personally wouldn't want someone else to take my work, completely rewrite it, and then still put me as "author". It's nice to mention that you have based your work on someone else's, but a Co-developed-by IMO implies that the other person was really involved in the patch.

I would definitely give a link to the downstream sources you have used as a base. But figuring out everyone who has ever made changes to `sdm660.dtsi` will be hard. Maybe the pinctrl driver was a slightly different situation because opendata largely copied it as-is (without modification).

minlexx commented:
In the end V3 of gcc-sdm660 [patch was applied](https://github.com/torvalds/linux/commit/f2a76a2955c0eb7514cdb5885e3d60a973301ae0) with folowing tags:

```
From: Stephen Boyd Oct. 16, 2018, 10:05 p.m. UTC
--------------------------------------------------------------------

Quoting Craig Tatlor (2018-09-25 10:35:58)
> From: Taniya Das <tdas@codeaurora.org>
> 
> Add support for the global clock controller found on SDM660
> based devices. This should allow most non-multimedia device
> drivers to probe and control their clocks.
> Based on CAF implementation.
> 
> Signed-off-by: Taniya Das <tdas@codeaurora.org>
> [craig: rename parents to fit upstream, and other cleanups]
> Signed-off-by: Craig Tatlor <ctatlor97@gmail.com>
> ---

Applied to clk-next
```

So I guess proceeding only with Signed-off-by should be fine for me...

minlexx committed:
dt-bindings: Add vendor prefix for Xiaomi

Xiaomi Corporation is a Chinese electronics company.

Signed-off-by: Alexey Minnekhanov <alexey.min@gmail.com>

minlexx committed:
dt-bindings: arm: qcom: Add sdm630 and sdm660 SoCs

Add a SoC strings for the Qualcomm SDM630 and SDM660 SoCs.
Also document the new xiaomi,lavender compatible used in
a device tree binding.

Signed-off-by: Alexey Minnekhanov <alexey.min@gmail.com>

minlexx commented:
Github does not display the correct order of commits when they are reordered in a MR; correct order can be seen in the branch https://github.com/minlexx/android_kernel_xiaomi_lavender/commits/qcom-sdm660-upstreaming

Changes:
* removed Linux Foundation from license headers in dts/dtsi files
* new commit adding xiaomi vendor prefix
* new commit adding new soc values for compatibles

stephan-gh reviewed (approved):
Looks good to me :+1: (of course this does not really mean anything for upstream, I'm not an expert either). But I think many possible comments have been avoided :)

minlexx commented:
I had to add
```patch
... 
   The 'board' element must be one of the following strings:
@@ -125,6 +127,11 @@ properties:
               - samsung,a5u-eur
           - const: qcom,msm8916
 
+      - items:
+          - enum:
+              - xiaomi,lavender
+          - const: qcom,sdm660
+
       - items:
           - const: qcom,msm8996-mtp
 
-- 
2.20.1
```
to the 3rd commit to fix the `WARNING: DT compatible string "xiaomi,lavender" appears un-documented`
